src/common/column_matrix.h

```diff
@@ -103,96 +110,179 @@ class ColumnMatrix {
+      #pragma omp parallel for num_threads(nthread)
+      for (omp_ulong rid = 0; rid < nrow; ++rid) {
+        const size_t ibegin = rid*nfeature;
+        const size_t iend = (rid+1)*nfeature;
+        size_t j = 0;
+        for (size_t i = ibegin; i < iend; ++i, ++j) {
+            const size_t idx = boundary_[j].index_begin;
+            local_index[idx + rid] = index[i];
+        }
+      }
+    } else {
+      size_t rbegin = 0;
+      for (const auto &batch : gmat.p_fmat_->GetBatches<SparsePage>()) {
+        for (size_t rid = 0; rid < batch.Size(); ++rid) {
+          SparsePage::Inst inst = batch[rid];
+          const size_t ibegin = gmat.row_ptr[rbegin + rid];
+          const size_t iend = gmat.row_ptr[rbegin + rid + 1];
+          CHECK_EQ(ibegin + inst.size(), iend);
+          size_t j = 0;
+          size_t fid = 0;
+          for (size_t i = ibegin; i < iend; ++i, ++j) {
+              fid = inst[j].index;
+              const size_t idx = boundary_[fid].index_begin;
+              local_index[idx + rbegin + rid] = index[i];
+              missing_flags_[idx + rbegin + rid] = false;
+          }
+        }
+        rbegin += batch.Size();
       }
     }
+  }
 
-    // loop over all rows and fill column entries
-    // num_nonzeros[fid] = how many nonzeros have this feature accumulated so far?
+  template<typename T>
+  inline void SetIndex(uint32_t* index, const GHistIndexMatrix& gmat,
+                       const size_t nrow, const size_t nfeature) {
     std::vector<size_t> num_nonzeros;
     num_nonzeros.resize(nfeature);
     std::fill(num_nonzeros.begin(), num_nonzeros.end(), 0);
-    for (size_t rid = 0; rid < nrow; ++rid) {
-      const size_t ibegin = gmat.row_ptr[rid];
-      const size_t iend = gmat.row_ptr[rid + 1];
-      size_t fid = 0;
-      for (size_t i = ibegin; i < iend; ++i) {
-        const uint32_t bin_id = gmat.index[i];
-        auto iter = std::upper_bound(gmat.cut.Ptrs().cbegin() + fid,
-                                     gmat.cut.Ptrs().cend(), bin_id);
-        fid = std::distance(gmat.cut.Ptrs().cbegin(), iter) - 1;
-        if (type_[fid] == kDenseColumn) {
-          uint32_t* begin = &index_[boundary_[fid].index_begin];
-          begin[rid] = bin_id - index_base_[fid];
-        } else {
-          uint32_t* begin = &index_[boundary_[fid].index_begin];
-          begin[num_nonzeros[fid]] = bin_id - index_base_[fid];
-          row_ind_[boundary_[fid].row_ind_begin + num_nonzeros[fid]] = rid;
-          ++num_nonzeros[fid];
+
+    T* local_index = reinterpret_cast<T*>(&index_[0]);
+
+    size_t rbegin = 0;
+    for (const auto &batch : gmat.p_fmat_->GetBatches<SparsePage>()) {
+      for (size_t rid = 0; rid < batch.Size(); ++rid) {
+        const size_t ibegin = gmat.row_ptr[rbegin + rid];
+        const size_t iend = gmat.row_ptr[rbegin + rid + 1];
+        size_t fid = 0;
+        SparsePage::Inst inst = batch[rid];
+        CHECK_EQ(ibegin + inst.size(), iend);
+        size_t jp = 0;
```

Comment:
What does `jp` mean?

Reply:
`j` variable is needed to get right feature index from `SparsePage` instance.
I think this check `CHECK_EQ(ibegin + inst.size(), iend);` should add some clarification.